pkg/node-controller/controller.go

```diff
@@ -0,0 +1,676 @@
+		// Foo resource to be synced.
+		if err := c.syncHandler(ctx, key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		logger.Info("Successfully synced", "resourceName", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Foo resource
+// with the current status of the resource.
+func (c *Controller) syncHandler(ctx context.Context, key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	logger := klog.LoggerWithValues(klog.FromContext(ctx), "resourceName", key)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	nad, err := c.nadLister.NetworkAttachmentDefinitions(namespace).Get(name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		// in this case the nad dne so it must've been deleted so we will cleanup nodeslicepools
+		// if we are down during the delete this could be missed similar to endpoints see kubernetes #6877
+		//TODO: do we want to cleanup IP Pools?
+		//err = c.whereaboutsclientset.WhereaboutsV1alpha1().NodeSlicePools(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+		//if err != nil && !errors.IsNotFound(err) {
+		//	return err
+		//}
+		//
+		//iPPools, err := c.ipPoolLister.IPPools(namespace).List(nil)
+		//if err != nil {
+		//	return err
+		//}
+		//for _, slicepool := range iPPools {
+		//	if ownerRef := metav1.GetControllerOf(slicepool); ownerRef != nil {
+		//		// If this object is not owned by a NodeSlicePool, we should not do anything more
+		//		// with it.
+		//		if ownerRef.Kind != "NodeSlicePool" || ownerRef.Name == name {
+		//			continue
+		//		}
+		//
+		//		err = c.whereaboutsclientset.WhereaboutsV1alpha1().IPPools(namespace).Delete(ctx, name, metav1.DeleteOptions{})
+		//		if err != nil {
+		//			if !errors.IsNotFound(err) {
+		//				return err
+		//			}
+		//		}
+		//	}
+		//}
+		return nil
+	}
+	//nad does exist so did it change node_slice_range or slice_size
+	//TODO: error syncing 'default/igor-1': config file not found, requeuing
+	// can we make this work without a local config file
+	ipamConf, err := ipamConfiguration(nad, "")
+	if err != nil {
+		return err
+	}
+
+	// This is to support several NADs and interfaces on the same network
+	//TODO: validation that nad configs are compatible between different nads with same network-name
+	sliceName := name
```

Comment:
this is an important feature for us, not sure if we could make it work nicely with overlapping ranges. We want to support many interfaces within the same network so many NADs will need to lock across the same node slices. This should also support "regular" NAD configs

this way it forces multiple nads to the same node slice allocations and then the same lease and IP Pool